File: src/services/stepfunctions.js
// services/stepfunctions.js — AWS Step Functions emulator
import { store, randomId, arn } from '../store.js';
import { jsonResponse, errorJson } from '../middleware/response.js';

function parseBody(req) {
  try { return JSON.parse(req.rawBody || '{}'); } catch { return {}; }
}

const TARGET_MAP = {
  'AWSStepFunctions.CreateStateMachine':   createStateMachine,
  'AWSStepFunctions.DeleteStateMachine':   deleteStateMachine,
  'AWSStepFunctions.DescribeStateMachine': describeStateMachine,
  'AWSStepFunctions.ListStateMachines':    listStateMachines,
  'AWSStepFunctions.UpdateStateMachine':   updateStateMachine,
  'AWSStepFunctions.StartExecution':       startExecution,
  'AWSStepFunctions.StopExecution':        stopExecution,
  'AWSStepFunctions.DescribeExecution':    describeExecution,
  'AWSStepFunctions.ListExecutions':       listExecutions,
  'AWSStepFunctions.GetExecutionHistory':  getExecutionHistory,
  'AWSStepFunctions.TagResource':          (req, res) => jsonResponse(res, 200, {}),
  'AWSStepFunctions.ListTagsForResource':  (req, res) => jsonResponse(res, 200, { tags: [] }),
};

export function handler(req, res) {
  const target = req.headers['x-amz-target'] || '';
  const fn = TARGET_MAP[target];
  if (fn) return fn(req, res);
  return errorJson(res, 400, 'InvalidAction', `Unknown Step Functions action: ${target}`);
}

function createStateMachine(req, res) {
  const b = parseBody(req);
  if (!b.name) return errorJson(res, 400, 'ValidationException', 'name required');
  if (store.stepfunctions.stateMachines[b.name])
    return errorJson(res, 400, 'StateMachineAlreadyExists', `State machine ${b.name} already exists`);
  const smArn = arn('states', `stateMachine:${b.name}`);
  store.stepfunctions.stateMachines[b.name] = {
    name:       b.name,
    arn:        smArn,
    definition: b.definition || '{}',
    roleArn:    b.roleArn || arn('iam', 'role/StatesRole'),
    type:       b.type || 'STANDARD',
    status:     'ACTIVE',
    created:    Date.now(),
    executions: [],
  };
  store.addTrail({ method: 'POST', path: `/states/${b.name}`, status: 200, latency: 5 });
  jsonResponse(res, 200, { stateMachineArn: smArn, creationDate: Date.now() / 1000 });
}

function deleteStateMachine(req, res) {
  const b = parseBody(req);
  const name = b.stateMachineArn?.split(':').pop();
  delete store.stepfunctions.stateMachines[name];
  jsonResponse(res, 200, {});
}

function describeStateMachine(req, res) {
  const b    = parseBody(req);
  const name = b.stateMachineArn?.split(':').pop();
  const sm   = store.stepfunctions.stateMachines[name];
  if (!sm) return errorJson(res, 400, 'StateMachineDoesNotExist', `State machine not found`);
  jsonResponse(res, 200, { ...sm, creationDate: sm.created / 1000 });
}

function listStateMachines(req, res) {
  jsonResponse(res, 200, {
    stateMachines: Object.values(store.stepfunctions.stateMachines).map(sm => ({
      name: sm.name, stateMachineArn: sm.arn, type: sm.type,
      creationDate: sm.created / 1000, status: sm.status,
    })),
  });
}

function updateStateMachine(req, res) {
  const b    = parseBody(req);
  const name = b.stateMachineArn?.split(':').pop();
  const sm   = store.stepfunctions.stateMachines[name];
  if (!sm) return errorJson(res, 400, 'StateMachineDoesNotExist', 'State machine not found');
  if (b.definition) sm.definition = b.definition;
  if (b.roleArn)    sm.roleArn    = b.roleArn;
  jsonResponse(res, 200, { updateDate: Date.now() / 1000 });
}

async function startExecution(req, res) {
  const b = parseBody(req);
  const name = b.stateMachineArn?.split(':').pop();
  const sm   = store.stepfunctions.stateMachines[name];
  if (!sm) return errorJson(res, 400, 'StateMachineDoesNotExist', 'State machine not found');

  const execName  = b.name || `exec-${randomId(8)}`;
  const execArn   = arn('states', `execution:${name}:${execName}`);
  const execution = {
    name:             execName,
    executionArn:     execArn,
    stateMachineArn:  sm.arn,
    input:            b.input || '{}',
    status:           'RUNNING',
    startDate:        Date.now() / 1000,
    stopDate:         null,
    output:           null,
    history:          [
      { timestamp: Date.now() / 1000, type: 'ExecutionStarted', executionStartedEventDetails: { input: b.input || '{}' } },
    ],
  };

  sm.executions.unshift(execution);
  store.stepfunctions.executions[execArn] = execution;
  store.addTrail({ method: 'POST', path: `/states/${name}/start`, status: 200, latency: 10 });

  // Simulate execution completing after 500ms
  setTimeout(() => {
    execution.status   = 'SUCCEEDED';
    execution.stopDate = Date.now() / 1000;
    execution.output   = b.input || '{}';
    execution.history.push({ timestamp: Date.now() / 1000, type: 'ExecutionSucceeded', executionSucceededEventDetails: { output: execution.output } });
  }, 500 + Math.random() * 1000);

  jsonResponse(res, 200, { executionArn: execArn, startDate: execution.startDate });
}

function stopExecution(req, res) {
  const b   = parseBody(req);
  const exec = store.stepfunctions.executions[b.executionArn];
  if (exec) { exec.status = 'ABORTED'; exec.stopDate = Date.now() / 1000; }
  jsonResponse(res, 200, { stopDate: Date.now() / 1000 });
}

function describeExecution(req, res) {
  const b    = parseBody(req);
  const exec = store.stepfunctions.executions[b.executionArn];
  if (!exec) return errorJson(res, 400, 'ExecutionDoesNotExist', 'Execution not found');
  jsonResponse(res, 200, exec);
}

function listExecutions(req, res) {
  const b    = parseBody(req);
  const name = b.stateMachineArn?.split(':').pop();
  const sm   = store.stepfunctions.stateMachines[name];
  if (!sm) return errorJson(res, 400, 'StateMachineDoesNotExist', 'State machine not found');
  const execs = sm.executions
    .filter(e => !b.statusFilter || e.status === b.statusFilter)
    .map(e => ({ name: e.name, executionArn: e.executionArn, stateMachineArn: e.stateMachineArn, status: e.status, startDate: e.startDate, stopDate: e.stopDate }));
  jsonResponse(res, 200, { executions: execs });
}

function getExecutionHistory(req, res) {
  const b    = parseBody(req);
  const exec = store.stepfunctions.executions[b.executionArn];
  if (!exec) return errorJson(res, 400, 'ExecutionDoesNotExist', 'Execution not found');
  jsonResponse(res, 200, { events: exec.history.map((e, i) => ({ id: i + 1, ...e })) });
}

File: internal/services/stepfunctions/stepfunctions.go
// Package stepfunctions — port of src/services/stepfunctions.js (JSON
// protocol, X-Amz-Target: AWSStepFunctions.*).
//
// The definition is kept as an UNPARSED value — there is no ASL interpreter
// (parity says keep the stub): StartExecution simulates completion after
// ~500–1500ms, echoing the input as the output. State lives in the any-tree
// state.SFNState (stateMachines + a global executions map that
// DeleteStateMachine purges in lockstep).
package stepfunctions

import (
	"encoding/json"
	"math/rand"
	"net/http"
	"sort"
	"strings"
	"time"

	"github.com/mockcloud/mockcloud/internal/httpapi"
	"github.com/mockcloud/mockcloud/internal/jsnum"
	"github.com/mockcloud/mockcloud/internal/protocol/respond"
	"github.com/mockcloud/mockcloud/internal/state"
	"github.com/mockcloud/mockcloud/internal/store"
)

type Service struct {
	st *store.Store
}

func New(st *store.Store) *Service { return &Service{st: st} }

// nameFromArn — Node's `arn?.split(':').pop()` (last segment; "" stays "").
func nameFromArn(arn string) string {
	parts := strings.Split(arn, ":")
	return parts[len(parts)-1]
}

// numF reads a numeric any-tree field (int64 when we stored it, float64 or
// json.Number after a snapshot import).
func numF(v any) float64 {
	switch t := v.(type) {
	case int64:
		return float64(t)
	case float64:
		return t
	case json.Number:
		f, _ := t.Float64()
		return f
	}
	return 0
}

func (svc *Service) Handler(w http.ResponseWriter, r *httpapi.Request) {
	target := r.Header.Get("x-amz-target")
	op := strings.TrimPrefix(target, "AWSStepFunctions.")
	if op == target {
		respond.ErrorJSON(w, 400, "InvalidAction", "Unknown Step Functions action: "+target)
		return
	}
	b := r.JSONBody()

	switch op {
	case "CreateStateMachine":
		svc.createStateMachine(w, b)
	case "DeleteStateMachine":
		svc.deleteStateMachine(w, b)
	case "DescribeStateMachine":
		svc.describeStateMachine(w, b)
	case "ListStateMachines":
		svc.listStateMachines(w)
	case "UpdateStateMachine":
		svc.updateStateMachine(w, b)
	case "StartExecution":
		svc.startExecution(w, b)
	case "StopExecution":
		svc.stopExecution(w, b)
	case "DescribeExecution":
		svc.describeExecution(w, b)
	case "ListExecutions":
		svc.listExecutions(w, b)
	case "GetExecutionHistory":
		svc.getExecutionHistory(w, b)
	case "TagResource":
		respond.JSON(w, 200, map[string]any{})
	case "ListTagsForResource":
		respond.JSON(w, 200, map[string]any{"tags": []any{}})
	default:
		respond.ErrorJSON(w, 400, "InvalidAction", "Unknown Step Functions action: "+target)
	}
}

// orDefault — JS `v || def` over any-tree values.
func orDefault(v, def any) any {
	if jsnum.Truthy(v) {
		return v
	}
	return def
}

func (svc *Service) createStateMachine(w http.ResponseWriter, b map[string]any) {
	name := httpapi.Str(b, "name")
	if name == "" {
		respond.ErrorJSON(w, 400, "ValidationException", "name required")
		return
	}
	smArn := state.Arn("states", "stateMachine:"+name)
	var exists bool
	svc.st.With(func(s *state.State) {
		if _, ok := s.StepFunctions.StateMachines[name]; ok {
			exists = true
			return
		}
		s.StepFunctions.StateMachines[name] = map[string]any{
			"name":       name,
			"arn":        smArn,
			"definition": orDefault(b["definition"], "{}"),
			"roleArn":    orDefault(b["roleArn"], state.IamArn("role/StatesRole")),
			"type":       orDefault(b["type"], "STANDARD"),
			"status":     "ACTIVE",
			"created":    state.NowMs(),
			"executions": []any{},
		}
		s.AddTrail(map[string]any{"method": "POST", "path": "/states/" + name, "status": 200, "latency": 5})
	})
	if exists {
		respond.ErrorJSON(w, 400, "StateMachineAlreadyExists", "State machine "+name+" already exists")
		return
	}
	respond.JSON(w, 200, map[string]any{"stateMachineArn": smArn, "creationDate": float64(state.NowMs()) / 1000})
}

func (svc *Service) deleteStateMachine(w http.ResponseWriter, b map[string]any) {
	name := nameFromArn(httpapi.Str(b, "stateMachineArn"))
	svc.st.With(func(s *state.State) {
		// Purge the machine's executions from the global map too, or they'd
		// be orphaned there forever (Node audit fix).
		if sm, ok := s.StepFunctions.StateMachines[name].(map[string]any); ok {
			if execs, ok := sm["executions"].([]any); ok {
				for _, e := range execs {
					if em, ok := e.(map[string]any); ok {
						if arn, ok := em["executionArn"].(string); ok {
							delete(s.StepFunctions.Executions, arn)
						}
					}
				}
			}
		}
		delete(s.StepFunctions.StateMachines, name)
	})
	respond.JSON(w, 200, map[string]any{})
}

func (svc *Service) describeStateMachine(w http.ResponseWriter, b map[string]any) {
	name := nameFromArn(httpapi.Str(b, "stateMachineArn"))
	var raw []byte
	svc.st.With(func(s *state.State) {
		sm, _ := s.StepFunctions.StateMachines[name].(map[string]any)
		if sm == nil {
			return
		}
		out := make(map[string]any, len(sm)+1)
		for k, v := range sm {
			out[k] = v
		}
		out["creationDate"] = numF(sm["created"]) / 1000
		raw = respond.Marshal(out) // marshal under the lock — timers mutate executions
	})
	if raw == nil {
		respond.ErrorJSON(w, 400, "StateMachineDoesNotExist", "State machine not found")
		return
	}
	respond.JSON(w, 200, json.RawMessage(raw))
}

func (svc *Service) listStateMachines(w http.ResponseWriter) {
	out := []map[string]any{}
	svc.st.With(func(s *state.State) {
		names := make([]string, 0, len(s.StepFunctions.StateMachines))
		for n := range s.StepFunctions.StateMachines {
			names = append(names, n)
		}
		sort.Strings(names)
		for _, n := range names {
			sm, _ := s.StepFunctions.StateMachines[n].(map[string]any)
			if sm == nil {
				continue
			}
			out = append(out, map[string]any{
				"name": sm["name"], "stateMachineArn": sm["arn"], "type": sm["type"],
				"creationDate": numF(sm["created"]) / 1000, "status": sm["status"],
			})
		}
	})
	respond.JSON(w, 200, map[string]any{"stateMachines": out})
}

func (svc *Service) updateStateMachine(w http.ResponseWriter, b map[string]any) {
	name := nameFromArn(httpapi.Str(b, "stateMachineArn"))
	var found bool
	svc.st.With(func(s *state.State) {
		sm, _ := s.StepFunctions.StateMachines[name].(map[string]any)
		if sm == nil {
			return
		}
		found = true
		if jsnum.Truthy(b["definition"]) {
			sm["definition"] = b["definition"]
		}
		if jsnum.Truthy(b["roleArn"]) {
			sm["roleArn"] = b["roleArn"]
		}
	})
	if !found {
		respond.ErrorJSON(w, 400, "StateMachineDoesNotExist", "State machine not found")
		return
	}
	respond.JSON(w, 200, map[string]any{"updateDate": float64(state.NowMs()) / 1000})
}

func (svc *Service) startExecution(w http.ResponseWriter, b map[string]any) {
	// Node: b.input || '{}' — and startStateMachineExecution stringifies
	// non-string inputs.
	input := "{}"
	if raw := b["input"]; jsnum.Truthy(raw) {
		if s, ok := raw.(string); ok {
			input = s
		} else {
			input = string(respond.Marshal(raw))
		}
	}
	exec := svc.StartStateMachineExecution(httpapi.Str(b, "stateMachineArn"), input, httpapi.Str(b, "name"))
	if exec == nil {
		respond.ErrorJSON(w, 400, "StateMachineDoesNotExist", "State machine not found")
		return
	}
	respond.JSON(w, 200, map[string]any{"executionArn": exec.ExecutionArn, "startDate": exec.StartDate})
}

// StartedExecution is the caller-visible slice of a new execution.
type StartedExecution struct {
	ExecutionArn string
	StartDate    float64
}

// StartStateMachineExecution — programmatic StartExecution, shared by the
// HTTP API and by EventBridge targets whose ARN is a state machine. Returns
// nil when the state machine doesn't exist. Must be called OUTSIDE the store
// lock (it takes it).
func (svc *Service) StartStateMachineExecution(stateMachineArn, input, execName string) *StartedExecution {
	name := nameFromArn(stateMachineArn)
	finalName := execName
	if finalName == "" {
		finalName = "exec-" + state.RandomID(8)
	}
	execArn := state.Arn("states", "execution:"+name+":"+finalName)

	var started *StartedExecution
	var exec map[string]any
	svc.st.With(func(s *state.State) {
		sm, _ := s.StepFunctions.StateMachines[name].(map[string]any)
		if sm == nil {
			return
		}
		now := float64(state.NowMs()) / 1000
		exec = map[string]any{
			"name":            finalName,
			"executionArn":    execArn,
			"stateMachineArn": sm["arn"],
			"input":           input,
			"status":          "RUNNING",
			"startDate":       now,
			"stopDate":        nil,
			"output":          nil,
			"history": []any{map[string]any{
				"timestamp": now, "type": "ExecutionStarted",
				"executionStartedEventDetails": map[string]any{"input": input},
			}},
		}
		execs, _ := sm["executions"].([]any)
		execs = append([]any{exec}, execs...)
		// Cap per-machine history, evicting from the global map in lockstep so
		// a recurring trigger can't grow either unboundedly. Loop, not if:
		// over-cap states from a snapshot import drain too.
		for len(execs) > 1000 {
			if em, ok := execs[len(execs)-1].(map[string]any); ok {
				if arn, ok := em["executionArn"].(string); ok {
					delete(s.StepFunctions.Executions, arn)
				}
			}
			execs = execs[:len(execs)-1]
		}
		sm["executions"] = execs
		s.StepFunctions.Executions[execArn] = exec
		s.AddTrail(map[string]any{"method": "POST", "path": "/states/" + name + "/start", "status": 200, "latency": 10})
		started = &StartedExecution{ExecutionArn: execArn, StartDate: now}
	})
	if started == nil {
		return nil
	}

	// Simulate execution completing after ~500–1500ms (Node's unref'd
	// setTimeout). Unconditional SUCCEEDED — even over an ABORTED status —
	// is Node behavior; after a store reset the map is unreachable, so the
	// late mutation is harmless there too.
	delay := time.Duration(500+rand.Float64()*1000) * time.Millisecond
	time.AfterFunc(delay, func() {
		svc.st.With(func(s *state.State) {
			now := float64(state.NowMs()) / 1000
			exec["status"] = "SUCCEEDED"
			exec["stopDate"] = now
			exec["output"] = input
			hist, _ := exec["history"].([]any)
			exec["history"] = append(hist, map[string]any{
				"timestamp": now, "type": "ExecutionSucceeded",
				"executionSucceededEventDetails": map[string]any{"output": input},
			})
		})
	})
	return started
}

func (svc *Service) stopExecution(w http.ResponseWriter, b map[string]any) {
	arn := httpapi.Str(b, "executionArn")
	svc.st.With(func(s *state.State) {
		if exec, ok := s.StepFunctions.Executions[arn].(map[string]any); ok {
			exec["status"] = "ABORTED"
			exec["stopDate"] = float64(state.NowMs()) / 1000
		}
	})
	respond.JSON(w, 200, map[string]any{"stopDate": float64(state.NowMs()) / 1000})
}

func (svc *Service) describeExecution(w http.ResponseWriter, b map[string]any) {
	arn := httpapi.Str(b, "executionArn")
	var raw []byte
	svc.st.With(func(s *state.State) {
		if exec, ok := s.StepFunctions.Executions[arn].(map[string]any); ok {
			raw = respond.Marshal(exec) // under the lock — the timer mutates it
		}
	})
	if raw == nil {
		respond.ErrorJSON(w, 400, "ExecutionDoesNotExist", "Execution not found")
		return
	}
	respond.JSON(w, 200, json.RawMessage(raw))
}

func (svc *Service) listExecutions(w http.ResponseWriter, b map[string]any) {
	name := nameFromArn(httpapi.Str(b, "stateMachineArn"))
	statusFilter := httpapi.Str(b, "statusFilter")
	var found bool
	out := []map[string]any{}
	svc.st.With(func(s *state.State) {
		sm, _ := s.StepFunctions.StateMachines[name].(map[string]any)
		if sm == nil {
			return
		}
		found = true
		execs, _ := sm["executions"].([]any)
		for _, e := range execs {
			em, ok := e.(map[string]any)
			if !ok {
				continue
			}
			if statusFilter != "" && em["status"] != statusFilter {
				continue
			}
			out = append(out, map[string]any{
				"name": em["name"], "executionArn": em["executionArn"],
				"stateMachineArn": em["stateMachineArn"], "status": em["status"],
				"startDate": em["startDate"], "stopDate": em["stopDate"],
			})
		}
	})
	if !found {
		respond.ErrorJSON(w, 400, "StateMachineDoesNotExist", "State machine not found")
		return
	}
	respond.JSON(w, 200, map[string]any{"executions": out})
}

func (svc *Service) getExecutionHistory(w http.ResponseWriter, b map[string]any) {
	arn := httpapi.Str(b, "executionArn")
	var raw []byte
	svc.st.With(func(s *state.State) {
		exec, ok := s.StepFunctions.Executions[arn].(map[string]any)
		if !ok {
			return
		}
		hist, _ := exec["history"].([]any)
		events := make([]map[string]any, 0, len(hist))
		for i, e := range hist {
			out := map[string]any{"id": i + 1}
			if em, ok := e.(map[string]any); ok {
				for k, v := range em {
					out[k] = v
				}
			}
			events = append(events, out)
		}
		raw = respond.Marshal(map[string]any{"events": events})
	})
	if raw == nil {
		respond.ErrorJSON(w, 400, "ExecutionDoesNotExist", "Execution not found")
		return
	}
	respond.JSON(w, 200, json.RawMessage(raw))
}
